### PR TITLE
feat(gradient): RadialGradientBrush for iOS, Android & WASM

### PR DESCRIPTION
**Pull request by carldebilly**
# Feature
First implementation of WinUI 2.4's `RadialGradientBrush`

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Media/GradientBrushTests/GradientsPage.xaml [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+							<GradientStop Color="LightGray" Offset="1" />
+						</not_win:RadialGradientBrush>
+					</Ellipse.Stroke>
+				</Ellipse>
+			</StackPanel>
+			<TextBlock FontSize="15">Linear Gradient as background on a Rectangle + Border (relative + absolute)</TextBlock>
+			<StackPanel Orientation="Horizontal" Spacing="10">
+				<Rectangle Width="100" Height="100">
+					<Rectangle.Fill>
+						<LinearGradientBrush not_win:FallbackColor="Red">
```
Comment (davidjohnoliver) [inline src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Media/GradientBrushTests/GradientsPage.xaml:121]: I would have the `LinearGradientBrush` and `RadialGradientBrush` samples in separate pages.

File: src/Uno.UI/UI/Xaml/IFrameworkElementImplementation.macOS.tt [hunk 2]
```diff
@@ -926,25 +926,32 @@ namespace <#= mixin.NamespaceName #>
 				WantsLayer = true;
 				var brush = e.NewValue as SolidColorBrush;
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/IFrameworkElementImplementation.macOS.tt:927]: seems like this variable shouldn't be necessary now

File: src/Uno.UI/UI/Xaml/Media/BrushMappingMode.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+namespace Windows.UI.Xaml.Media
+{
+	public enum BrushMappingMode 
+	{
+		[Uno.NotImplemented]
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Media/BrushMappingMode.cs:5]: I don't really understand how 1 out of 2 `enum` values can be implemented, given that you have a conditional on `== RelativeToBoundingBox` below.

File: src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+		public Point GradientOrigin
+		{
+			get => (Point)GetValue(GradientOriginProperty);
+			set => SetValue(GradientOriginProperty, value);
+		}
+
+		public static readonly DependencyProperty InterpolationSpaceProperty = DependencyProperty.Register(
+			"InterpolationSpace", typeof(CompositionColorSpace), typeof(RadialGradientBrush), new PropertyMetadata(default(CompositionColorSpace)));
+
+		public CompositionColorSpace InterpolationSpace
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs:48]: Is this actually being used? Should it be marked `[NotImplemented]`?

File: src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
+﻿using Windows.Foundation;
+using Windows.UI.Composition;
+using Windows.UI.Xaml.Markup;
+
+namespace Windows.UI.Xaml.Media
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs:5]: Should this be the `Microsoft.UI.Xaml.Media` namespace? cc @jeromelaban

File: src/Uno.UI.Wasm/WasmCSS/Uno.UI.css [hunk 6]
```diff
@@ -99,6 +99,10 @@ embed.uno-frameworkelement.uno-unarranged {
   -moz-user-select: none; /* Firefox */
   -ms-user-select: none; /* Internet Explorer/Edge */
   user-select: none; /* Non-prefixed version, currently supported by Chrome and Opera */
+
+  -ms-background-clip: text !important;
+  -webkit-background-clip: text !important;
+  background-clip: text !important;
```
Comment (jeromelaban) [inline src/Uno.UI.Wasm/WasmCSS/Uno.UI.css:105]: Can you indicate the use of this change, so we know easily why is needed?

File: src/Uno.UI/Controls/BindableImageView.Android.cs [hunk 7]
```diff
@@ -270,6 +270,7 @@ private void OnUriSourceChanged()
 				var sampleSize = CalculateInSampleSize(options, _targetWidth, _targetHeight);
 
 				options.InJustDecodeBounds = false;
+#pragma warning disable CS0618 // Type or member is obsolete
```
Comment (jeromelaban) [inline src/Uno.UI/Controls/BindableImageView.Android.cs:273]: Restore the warning after `InPurgeable`

File: src/Uno.UI/UI/Xaml/IFrameworkElementImplementation.iOS.tt [hunk 8]
```diff
@@ -935,26 +935,37 @@ namespace <#= mixin.NamespaceName #>
...
+						GradientBrush.OpacityProperty,
+						(s, colorArg) => BackgroundColor = (s as GradientBrush).FallbackColorWithOpacity
+					);
+				}
 				else
 				{
-					this.Log().Warn("Brush is not a SolidColorBrush, ignoring");
 					_brushColorChanged.Disposable = null;
 					_brushOpacityChanged.Disposable = null;
+					this.Log().Warn("Brush is not a SolidColorBrush, ignoring");
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/IFrameworkElementImplementation.iOS.tt:968]: Add a if(Level) around

File: src/Uno.UI/UI/Xaml/IFrameworkElementImplementation.macOS.tt [hunk 9]
```diff
@@ -926,25 +926,32 @@ namespace <#= mixin.NamespaceName #>
...
 					_brushColorChanged.Disposable = null;
 					_brushOpacityChanged.Disposable = null;
-					Layer.BackgroundColor = SolidColorBrushHelper.Transparent.Color;
+					if(e.NewValue is GradientBrush gb)
+					{
+						Layer.BackgroundColor = gb.FallbackColorWithOpacity;
+					}
+					else
+					{
+						this.Log().Warn("Brush is not a SolidColorBrush, ignoring");
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/IFrameworkElementImplementation.macOS.tt:952]: Add an if(level) around

File: src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs (hunk 5 above)
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs:5]: Indeed:

https://github.com/microsoft/microsoft-ui-xaml/blob/ca95be6f5248eaa55ec523f2d6e7b50f6e0d53e7/dev/RadialGradientBrush/RadialGradientBrush.idl#L1

In WinUI 2.x, this type is used by qualifying explicitly in the xaml.

File: src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+			"Center", typeof(Point), typeof(RadialGradientBrush), new PropertyMetadata(new Point(0.5d, 0.5d)));
+
+		public Point Center
+		{
+			get => (Point)GetValue(CenterProperty);
+			set => SetValue(CenterProperty, value);
+		}
+
+		public static readonly DependencyProperty RadiusXProperty = DependencyProperty.Register(
+			"RadiusX", typeof(double), typeof(RadialGradientBrush), new PropertyMetadata(1.0d));
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs:19]: The default value in WinUI is `0.5`: https://github.com/microsoft/microsoft-ui-xaml/blob/ca95be6f5248eaa55ec523f2d6e7b50f6e0d53e7/dev/RadialGradientBrush/RadialGradientBrush.idl#L16

File: src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+		public double RadiusX
+		{
+			get => (double)GetValue(RadiusXProperty);
+			set => SetValue(RadiusXProperty, value);
+		}
+
+		public static readonly DependencyProperty RadiusYProperty = DependencyProperty.Register(
+			"RadiusY", typeof(double), typeof(RadialGradientBrush), new PropertyMetadata(1.0d));
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs:28]: Same here, it's `0.5`.

File: src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+		public double RadiusY
+		{
+			get => (double)GetValue(RadiusYProperty);
+			set => SetValue(RadiusYProperty, value);
+		}
+
+		public static readonly DependencyProperty GradientOriginProperty = DependencyProperty.Register(
+			"GradientOrigin", typeof(Point), typeof(RadialGradientBrush), new PropertyMetadata(default(Point)));
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs:37]:
```suggestion
			"GradientOrigin", typeof(Point), typeof(RadialGradientBrush), new PropertyMetadata(new Point(0.5, 0.5)));
```

https://github.com/microsoft/microsoft-ui-xaml/blob/ca95be6f5248eaa55ec523f2d6e7b50f6e0d53e7/dev/RadialGradientBrush/RadialGradientBrush.idl#L24

File: src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs [hunk 13]
```diff
@@ -0,0 +1,54 @@
+﻿using Windows.Foundation;
+using Windows.UI.Composition;
+using Windows.UI.Xaml.Markup;
+
+namespace Windows.UI.Xaml.Media
+{
+	public sealed partial class RadialGradientBrush : GradientBrush
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs:7]: Also the default value for `MappingMode` is `RelativeToBoundingBox`.

https://github.com/microsoft/microsoft-ui-xaml/blob/ca95be6f5248eaa55ec523f2d6e7b50f6e0d53e7/dev/RadialGradientBrush/RadialGradientBrush.idl#L28

File: src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.wasm.cs [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+		internal override UIElement ToSvgElement()
+		{
+			var center = Center;
+
+			var radiusX = RadiusX;
+			var radiusY = RadiusY;
+
+			if (MappingMode != BrushMappingMode.RelativeToBoundingBox)
+			{
+				// TODO: not supported right now
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.wasm.cs:41]: Probably should raise a message here with `ApiInformation`

File: src/Uno.UI/Controls/BindableImageView.Android.cs (hunk 7 above)
Comment (carldebilly) [reply src/Uno.UI/Controls/BindableImageView.Android.cs:273]: Wasn't supposed to be in the commit.  Will be removed.

File: src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs (hunk 4 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs:52]: No it's not yet...

File: src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs (hunk 13 above)
Comment (carldebilly) [reply src/Uno.UI/UI/Xaml/Media/RadialGradientBrush.cs:10]: It's already the case... Check in `GradientBrush.cs`.

**Pull request merged**